src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,296 @@
+
+            validationMessages.AddRange(semanticDiagnostics.Select(DiagnosticToValidationMessage));
+
+            return CreateAndSetValidationResult(out _syntaxValidationResult,
+                !ErrorExists(validationMessages),
+                validationMessages);
+        }
+
+        public Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var algoClassFullName = _syntaxWalker.NamespaceNode == null
+                ? $"{_syntaxWalker.ClassNode.Identifier}"
+                : $"{_syntaxWalker.NamespaceNode.Name}.{_syntaxWalker.ClassNode.Identifier}";
+
+            var newAssembly = GenerateAssembly();
+            var newType = newAssembly.GetType(algoClassFullName);
+            var metadata = ExtractMetadata(newType);
+
+            return Task.FromResult(metadata);
+        }
+
+        private AlgoMetaDataInformation ExtractMetadata(Type algoType)
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            //Get public properties that have public setter
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public)
+                .Where(x => x.GetSetMethod() != null);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                //Base parameters
+                if (algoProperty.PropertyType.BaseType != typeof(AbstractFunction)
+                    && !typeof(IFunction).IsAssignableFrom(algoProperty.PropertyType))
+                {
+                    var parameter = ToAlgoMetaDataParameter(algoProperty);
+
+                    if (algoProperty.PropertyType.IsEnum)
+                        parameter.PredefinedValues = ToEnumValues(algoProperty);
+
+                    metadata.Parameters.Add(parameter);
+                }
+                //Functions
+                else if (algoProperty.PropertyType.BaseType == typeof(AbstractFunction)
+                         || typeof(IFunction).IsAssignableFrom(algoProperty.PropertyType))
+                {
+                    //Get first public property that is not function base parameters but it inherits it
+                    var functionProperty = algoProperty.PropertyType
+                        .GetProperties(BindingFlags.Instance | BindingFlags.Public | BindingFlags.DeclaredOnly |
+                                       BindingFlags.GetProperty |
+                                       BindingFlags.SetProperty) //BindingFlags.DeclaredOnly -> ignore inherited members
+                        .FirstOrDefault(x => x.GetSetMethod() != null &&
+                                             typeof(FunctionParamsBase).IsAssignableFrom(x.PropertyType));
+
+                    //If there is no such property, get property that is function base parameters
+                    if (functionProperty == null)
```

Comment:
There is no need for this, `typeof(FunctionParamsBase).IsAssignableFrom(x.PropertyType)` will be true even if the property is of type `FunctionParamsBase`